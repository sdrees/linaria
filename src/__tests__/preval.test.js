/* eslint-disable no-template-curly-in-string */
/* @flow */

const path = require('path');
const babel = require('@babel/core');
const dedent = require('dedent');
const stripAnsi = require('strip-ansi');
const serializer = require('../__utils__/linaria-snapshot-serializer');

expect.addSnapshotSerializer(serializer);

const babelrc = {
  babelrc: false,
  presets: [
    [require.resolve('../babel'), { displayName: true, evaluate: true }],
  ],
};

const transpile = async input => {
  const { code, metadata } = await babel.transformAsync(input, {
    ...babelrc,
    filename: path.join(__dirname, 'source.js'),
  });

  // The slug will be machine specific, so replace it with a consistent one
  return {
    metadata,
    code,
  };
};

it('evaluates identifier in scope', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const answer = 42;
    const foo = () => answer;
    const days = foo() + ' days';

    export const Title = styled.h1\`
      &:before {
        content: "${'${days}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates local expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const answer = 42;
    const foo = () => answer;

    export const Title = styled.h1\`
      &:before {
        content: "${"${foo() + ' days'}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates functions with nested identifiers', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const objects = { key: { fontSize: 12 } };
    const foo = (k) => {
      const obj = objects[k];
      return obj;
    };

    export const Title = styled.h1\`
      ${"${foo('key')}"}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates expressions with dependencies', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';
    import slugify from '../slugify';

    export const Title = styled.h1\`
      &:before {
        content: "${"${slugify('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates expressions with expressions depending on shared dependency', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';
    const slugify = require('../slugify');

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    export const Title = styled.h1\`
      &:before {
        content: "${"${boo('test') + bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates multiple expressions with shared dependency', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';
    const slugify = require('../slugify');

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    export const Title = styled.h1\`
      &:before {
        content: "${"${boo('test')}"}"
        content: "${"${bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evalutes interpolations with sequence expression', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    export const Title = styled.h1\`
      color: ${'${(external, () => "blue")}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evalutes dependencies with sequence expression', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const color = (external, () => 'blue');

    export const Title = styled.h1\`
      color: ${'${color}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates component interpolations', async () => {
  const { code, metadata } = await transpile(
    dedent`
    const { styled } = require('../react');

    export const Title = styled.h1\`
      color: red;
    \`;

    export const Paragraph = styled.p\`
      ${'${Title}'} {
        color: blue;
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws when interpolation evaluates to undefined', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      let fontSize;

      export const Title = styled.h1\`
        font-size: ${'${fontSize}'};
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('throws when interpolation evaluates to null', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      const color = null;

      export const Title = styled.h1\`
        color: ${'${color}'};
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('throws when interpolation evaluates to NaN', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      const height = NaN;

      export const Title = styled.h1\`
        height: ${'${height}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('handles wrapping another styled component', async () => {
  const { code, metadata } = await transpile(
    dedent`
    const { styled } = require('../react');

    const Title = styled.h1\`
      color: red;
    \`;

    export const CustomTitle = styled(Title)\`
      font-size: 24px;
      color: blue;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines object styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const fill = (top = 0, left = 0, right = 0, bottom = 0) => ({
      position: 'absolute',
      top,
      right,
      bottom,
      left,
    });

    export const Title = styled.h1\`
      ${'${fill(0, 0)}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines array styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const fill = (top = 0, left = 0, right = 0, bottom = 0) => [
      { position: 'absolute' },
      {
        top,
        right,
        bottom,
        left,
      }
    ];

    export const Title = styled.h1\`
      ${'${fill(0, 0)}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores inline arrow function expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    export const Title = styled.h1\`
      &:before {
        content: "${'${props => props.content}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores inline vanilla function expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    export const Title = styled.h1\`
      &:before {
        content: "${'${function(props) { return props.content }}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores external expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from 'linaria/react';

    const generate = props => props.content;

    export const Title = styled.h1\`
      &:before {
        content: "${'${generate}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws codeframe error when evaluation fails', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      import { styled } from 'linaria/react';

      const foo = props => { throw new Error('This will fail') };

      export const Title = styled.h1\`
        font-size: ${'${foo()}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('handles escapes properly', async () => {
  const { code, metadata } = await babel.transformFileAsync(
    path.resolve(__dirname, '../__fixtures__/escape-character.js'),
    babelrc
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('derives display name from filename', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from 'linaria/react';

    export default styled.h1\`
      font-size: 14px;
    \`;
    `,
    {
      ...babelrc,
      filename: path.join(__dirname, 'FancyName.js'),
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('derives display name from parent folder name', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from 'linaria/react';

    export default styled.h1\`
      font-size: 14px;
    \`;
    `,
    {
      ...babelrc,
      filename: path.join(__dirname, 'FancyName/index.js'),
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it("throws if couldn't determine a display name", async () => {
  expect.assertions(1);

  try {
    await babel.transformAsync(
      dedent`
      import { styled } from 'linaria/react';

      export default styled.h1\`
        font-size: 14px;
      \`;
      `,
      {
        ...babelrc,
        filename: path.join(__dirname, '/.js'),
      }
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('does not strip instanbul coverage sequences', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from 'linaria/react';

    const a = 42;

    export const Title = styled.h1\`
      height: ${'${a}'}px;
    \`;
    `,
    {
      ...babelrc,
      cwd: '/home/user/project',
      filename: 'file.js',
      plugins: [
        [
          // eslint-disable-next-line import/no-extraneous-dependencies
          require('babel-plugin-istanbul').default({ types: babel.types }),
          { cwd: '/home/user/project' },
        ],
      ],
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});
