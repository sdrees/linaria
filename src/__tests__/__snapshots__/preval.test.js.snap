// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`derives display name from filename 1`] = `
"import { styled } from 'linaria/react';
export default
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"FancyName0\\",
  class: \\"FancyName0_f1yjazsq\\"
});"
`;

exports[`derives display name from filename 2`] = `

CSS:

.FancyName0_f1yjazsq {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`derives display name from parent folder name 1`] = `
"import { styled } from 'linaria/react';
export default
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"FancyName0\\",
  class: \\"FancyName0_f1e8u5dq\\"
});"
`;

exports[`derives display name from parent folder name 2`] = `

CSS:

.FancyName0_f1e8u5dq {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`does not strip instanbul coverage sequences 1`] = `
"var cov_2dr9r1nt95 = function () {
  var path = \\"/home/user/project/file.js\\";
  var hash = \\"23f56bbaeb2ebcd213f41374d3b2af1bce287bd3\\";

  var Function = function () {}.constructor;

  var global = new Function(\\"return this\\")();
  var gcv = \\"__coverage__\\";
  var coverageData = {
    path: \\"/home/user/project/file.js\\",
    statementMap: {
      \\"0\\": {
        start: {
          line: 3,
          column: 10
        },
        end: {
          line: 3,
          column: 12
        }
      },
      \\"1\\": {
        start: {
          line: 5,
          column: 21
        },
        end: {
          line: 7,
          column: 1
        }
      }
    },
    fnMap: {},
    branchMap: {},
    s: {
      \\"0\\": 0,
      \\"1\\": 0
    },
    f: {},
    b: {},
    _coverageSchema: \\"43e27e138ebf9cfc5966b082cf9a028302ed4184\\"
  };
  var coverage = global[gcv] || (global[gcv] = {});

  if (coverage[path] && coverage[path].hash === hash) {
    return coverage[path];
  }

  coverageData.hash = hash;
  return coverage[path] = coverageData;
}();

import { styled } from 'linaria/react';
const a = (cov_2dr9r1nt95.s[0]++, 42);
export const Title = (cov_2dr9r1nt95.s[1]++,
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_tow9xsn\\"
}));"
`;

exports[`does not strip instanbul coverage sequences 2`] = `

CSS:

.Title_tow9xsn {
  height: 42px;
}

Dependencies: NA

`;

exports[`evaluates component interpolations 1`] = `
"const {
  styled
} = require('../react');

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});
export const Paragraph =
/*#__PURE__*/
styled(\\"p\\")({
  name: \\"Paragraph\\",
  class: \\"Paragraph_p1rsdnkv\\"
});"
`;

exports[`evaluates component interpolations 2`] = `

CSS:

.Title_t1xha7dm {
  color: red;
}
.Paragraph_p1rsdnkv {
  .Title_t1xha7dm {
    color: blue;
  }
}

Dependencies: ../react

`;

exports[`evaluates expressions with dependencies 1`] = `
"import { styled } from 'linaria/react';
import slugify from '../slugify';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates expressions with dependencies 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jy"
  }
}

Dependencies: ../slugify

`;

exports[`evaluates expressions with expressions depending on shared dependency 1`] = `
"import { styled } from 'linaria/react';

const slugify = require('../slugify');

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates expressions with expressions depending on shared dependency 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jyboo6og6jybar"
  }
}

Dependencies: ../slugify

`;

exports[`evaluates functions with nested identifiers 1`] = `
"import { styled } from 'linaria/react';
const objects = {
  key: {
    fontSize: 12
  }
};

const foo = k => {
  const obj = objects[k];
  return obj;
};

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates functions with nested identifiers 2`] = `

CSS:

.Title_t1xha7dm {
  font-size: 12px;
}

Dependencies: NA

`;

exports[`evaluates identifier in scope 1`] = `
"import { styled } from 'linaria/react';
const answer = 42;

const foo = () => answer;

const days = foo() + ' days';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates identifier in scope 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "42 days"
  }
}

Dependencies: NA

`;

exports[`evaluates local expressions 1`] = `
"import { styled } from 'linaria/react';
const answer = 42;

const foo = () => answer;

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates local expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "42 days"
  }
}

Dependencies: NA

`;

exports[`evaluates multiple expressions with shared dependency 1`] = `
"import { styled } from 'linaria/react';

const slugify = require('../slugify');

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates multiple expressions with shared dependency 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jyboo"
    content: "6og6jybar"
  }
}

Dependencies: ../slugify, ../slugify

`;

exports[`evalutes dependencies with sequence expression 1`] = `
"import { styled } from 'linaria/react';
const color = (external, () => 'blue');
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [color]
  }
});"
`;

exports[`evalutes dependencies with sequence expression 2`] = `

CSS:

.Title_t1xha7dm {
  color: var(--t1xha7dm-0);
}

Dependencies: NA

`;

exports[`evalutes interpolations with sequence expression 1`] = `
"import { styled } from 'linaria/react';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [(external, () => \\"blue\\")]
  }
});"
`;

exports[`evalutes interpolations with sequence expression 2`] = `

CSS:

.Title_t1xha7dm {
  color: var(--t1xha7dm-0);
}

Dependencies: NA

`;

exports[`handles escapes properly 1`] = `
"import { styled } from 'linaria/react';
const selectors = ['a', 'b'];
export const Block =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Block\\",
  class: \\"Block_b3rflbm\\"
});"
`;

exports[`handles escapes properly 2`] = `

CSS:

.Block_b3rflbm {
  a { content: "\\u000A"; } b { content: "\\u000A"; }



   ;
}

Dependencies: core-js/modules/es6.string.raw

`;

exports[`handles wrapping another styled component 1`] = `
"const {
  styled
} = require('../react');

const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});
export const CustomTitle =
/*#__PURE__*/
styled(Title)({
  name: \\"CustomTitle\\",
  class: \\"CustomTitle_c1rsdnkv\\"
});"
`;

exports[`handles wrapping another styled component 2`] = `

CSS:

.Title_t1xha7dm {
  color: red;
}
.CustomTitle_c1rsdnkv.Title_t1xha7dm {
  font-size: 24px;
  color: blue;
}

Dependencies: NA

`;

exports[`ignores external expressions 1`] = `
"import { styled } from 'linaria/react';

const generate = props => props.content;

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [generate]
  }
});"
`;

exports[`ignores external expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: NA

`;

exports[`ignores inline arrow function expressions 1`] = `
"import { styled } from 'linaria/react';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [props => props.content]
  }
});"
`;

exports[`ignores inline arrow function expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: NA

`;

exports[`ignores inline vanilla function expressions 1`] = `
"import { styled } from 'linaria/react';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [function (props) {
      return props.content;
    }]
  }
});"
`;

exports[`ignores inline vanilla function expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: NA

`;

exports[`inlines array styles as CSS string 1`] = `
"import { styled } from 'linaria/react';

const fill = (top = 0, left = 0, right = 0, bottom = 0) => [{
  position: 'absolute'
}, {
  top,
  right,
  bottom,
  left
}];

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`inlines array styles as CSS string 2`] = `

CSS:

.Title_t1xha7dm {
  position: absolute; top: 0; right: 0; bottom: 0; left: 0;
}

Dependencies: NA

`;

exports[`inlines object styles as CSS string 1`] = `
"import { styled } from 'linaria/react';

const fill = (top = 0, left = 0, right = 0, bottom = 0) => ({
  position: 'absolute',
  top,
  right,
  bottom,
  left
});

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`inlines object styles as CSS string 2`] = `

CSS:

.Title_t1xha7dm {
  position: absolute; top: 0; right: 0; bottom: 0; left: 0;
}

Dependencies: NA

`;

exports[`throws codeframe error when evaluation fails 1`] = `
"<<DIRNAME>>/source.js: An error occurred when evaluating the expression: This will fail. Make sure you are not using a browser or Node specific API.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   font-size: \${foo()}px;
    |                ^
  7 | \`;"
`;

exports[`throws if couldn't determine a display name 1`] = `
"<<DIRNAME>>/.js: Couldn't determine a name for the component. Ensure that it's either:
- Assigned to a variable
- Is an object property
- Is a prop in a JSX element

  1 | import { styled } from 'linaria/react';
  2 | 
> 3 | export default styled.h1\`
    |                ^
  4 |   font-size: 14px;
  5 | \`;"
`;

exports[`throws when interpolation evaluates to NaN 1`] = `
"<<DIRNAME>>/source.js: The expression evaluated to 'NaN', which is probably a mistake. If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(height)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   height: \${height}px;
    |             ^
  7 | \`;"
`;

exports[`throws when interpolation evaluates to null 1`] = `
"<<DIRNAME>>/source.js: The expression evaluated to 'null', which is probably a mistake. If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(color)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   color: \${color};
    |            ^
  7 | \`;"
`;

exports[`throws when interpolation evaluates to undefined 1`] = `
"<<DIRNAME>>/source.js: The expression evaluated to 'undefined', which is probably a mistake. If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(fontSize)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   font-size: \${fontSize};
    |                ^
  7 | \`;"
`;
